### PR TITLE
Add Cell Ranger mkfastq, mkgtf, and count

### LINKED ISSUE #926: new module: `cellranger/count`
I think it would be good to have a module for cellranger count

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [ ] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by Emiller88**
Adds `cellranger count` and needed `cellranger mkgtf` to create a filtered gtf

## PR checklist

Closes #926

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

We'll see if the conda tests pass? I assume the `mkref` one did.

### CODE REVIEW

**Review by lescai — APPROVED** [review]
Looks good to me!

File: modules/cellranger/Dockerfile [hunk 1]
```diff
@@ -4,7 +4,7 @@ LABEL authors="Gisela Gabernet <gisela.gabernet@gmail.com>" \
 # Disclaimer: this container is not provided nor supported by 10x Genomics.
 
 # Install procps and clean apt cache
```
Comment (drpatelh) [inline modules/cellranger/Dockerfile:6]: We need to decide how we are going to use these `Dockerfile` properly for these scenarios.

File: modules/cellranger/count/main.nf [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:[:], publish_by_meta:[]) }
+
+    if (params.enable_conda) {
+        exit 1, "Conda environments cannot be used when using the Cell Ranger tool. Please use docker or singularity containers."
+    }
+    container "nfcore/cellranger:6.0.2"
+
+    input:
+    tuple val(meta), path(reads)
+    path(reference)
```
Comment (drpatelh) [inline modules/cellranger/count/main.nf:21]:
```suggestion
    path reference
```

File: modules/cellranger/count/main.nf [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    def sample_arg = meta.samples.unique().join(",")
+    def reference_name = reference.name
+    """
+    cellranger count --id='sample-${meta.gem}' \
+        --fastqs=. \
+        --transcriptome=${reference_name} \
+        --sample=${sample_arg} \
+        --localcores=${task.cpus} \
+        --localmem=${task.memory.toGiga()} \
+        $options.args
```
Comment (drpatelh) [inline modules/cellranger/count/main.nf:37]:
```suggestion
    cellranger \\
        count \\
        --id='sample-${meta.gem}' \\
        --fastqs=. \\
        --transcriptome=$reference_name \\
        --sample=$sample_arg \\
        --localcores=$task.cpus \\
        --localmem=${task.memory.toGiga()} \\
        $options.args
```

File: modules/cellranger/count/main.nf [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    cat <<-END_VERSIONS > versions.yml
+    ${getProcessName(task.process)}:
+        ${getSoftwareName(task.process)}: \$(echo \$( cellranger --version 2>&1) | sed 's/^.*[^0-9]\\([0-9]*\\.[0-9]*\\.[0-9]*\\).*\$/\\1/' )
+    END_VERSIONS
+    """
+
+    stub:
```
Comment (drpatelh) [inline modules/cellranger/count/main.nf:45]: Noice!

File: modules/cellranger/mkgtf/main.nf [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        exit 1, "Conda environments cannot be used when using the Cell Ranger tool. Please use docker or singularity containers."
+    }
+    container "nfcore/cellranger:6.0.2"
+
+    input:
+    path gtf
+
+    output:
+    path "versions.yml"   , emit: versions
+    path "*.filtered.gtf" , emit: gtf
```
Comment (drpatelh) [inline modules/cellranger/mkgtf/main.nf:24]:
```suggestion
    path "*.filtered.gtf", emit: gtf
    path "versions.yml"  , emit: versions

```

File: modules/cellranger/mkgtf/main.nf [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+    script:
+    """
+    cellranger mkgtf \
+        $gtf \
+        ${gtf.baseName}.filtered.gtf \
+        $options.args \
+        $options.args2
```
Comment (drpatelh) [inline modules/cellranger/mkgtf/main.nf:32]:
```suggestion
    cellranger \\
        mkgtf \\
        $gtf \\
        ${gtf.baseName}.filtered.gtf \\
        $options.args \\
        $options.args2
```

File: modules/cellranger/mkgtf/meta.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      documentation: https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_ov
+      tool_dev_url: https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_ov
+      doi: ""
+      licence: 10x Genomics EULA
+
+input:
+  - fasta:
+      type: file
+      description: fasta genome file
+      pattern: "*.{fasta,fa}"
```
Comment (drpatelh) [inline modules/cellranger/mkgtf/meta.yml:20]:
```suggestion
```

File: modules/cellranger/mkgtf/meta.yml [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  - gtf:
+      type: file
+      description: gtf transcriptome file
+      pattern: "*.gtf"
+  - reference_name:
+      type: val
+      description: name to give the reference folder
+      pattern: str
```
Comment (drpatelh) [inline modules/cellranger/mkgtf/meta.yml:28]:
```suggestion
```

File: modules/cellranger/mkgtf/meta.yml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+output:
+  - versions:
+      type: file
+      description: File containing software version
+      pattern: "versions.yml"
+  - reference:
```
Comment (drpatelh) [inline modules/cellranger/mkgtf/meta.yml:35]:
```suggestion
  - gtf:
```

File: tests/config/test_data.config [hunk 10]
```diff
@@ -211,14 +211,8 @@ params {
...
-                test_broadpeak                                 = "${test_data_dir}/genomics/homo_sapiens/illumina/broadpeak/test.broadPeak"
-                test2_broadpeak                                = "${test_data_dir}/genomics/homo_sapiens/illumina/broadpeak/test2.broadPeak"
-
-                test_narrowpeak                                = "${test_data_dir}/genomics/homo_sapiens/illumina/narrowpeak/test.narrowPeak"
-                test2_narrowpeak                               = "${test_data_dir}/genomics/homo_sapiens/illumina/narrowpeak/test2.narrowPeak"
-
-                test_10x_1_fastq_gz                           = "${test_data_dir}/genomics/homo_sapiens/illumina/10xgenomics/test.10x_1.fastq.gz"
-                test_10x_2_fastq_gz                           = "${test_data_dir}/genomics/homo_sapiens/illumina/10xgenomics/test.10x_2.fastq.gz"
+                test_10x_1_fastq_gz                           = "${test_data_dir}/genomics/homo_sapiens/illumina/10xgenomics/test_10x_S1_L001_R1_001.fastq.gz"
+                test_10x_2_fastq_gz                           = "${test_data_dir}/genomics/homo_sapiens/illumina/10xgenomics/test_10x_S1_L001_R2_001.fastq.gz"
```
Comment (drpatelh) [inline tests/config/test_data.config:215]: Can we rename these files on nf-core/testdatasets to their original names i.e. `test.10x_1.fastq.gz` and `test.10x_2.fastq.gz`

File: modules/cellranger/.gitignore [hunk 11]
```diff
@@ -0,0 +1 @@
+cellranger-*.tar.gz
```
Comment (drpatelh) [inline modules/cellranger/.gitignore:1]: Why is this file here?

**Review by drpatelh — COMMENTED** [review]
Looks great @Emiller88 . Still some work to do to get everything consistent.

File: modules/cellranger/.gitignore (hunk 11 above)
Comment (Emiller88) [reply modules/cellranger/.gitignore:1]: When you build the docker image you have to put cell ranger in the same directory. It's a just in-case anyone likes to `git add .` and `git commit`

File: modules/cellranger/count/main.nf (hunk 4 above)
Comment (Emiller88) [reply modules/cellranger/count/main.nf:45]: Oh this was @ggabernet. Maybe this is what I should do for spaceranger though :thinking:

File: tests/config/test_data.config (hunk 10 above)
Comment (Emiller88) [reply tests/config/test_data.config:215]: Negative. Cell ranger only supports Illumina file naming for the moment, sadly. :disappointed:

File: modules/cellranger/mkgtf/main.nf [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    container "nfcore/cellranger:6.0.2"
+
+    input:
+    path gtf
+
+    output:
+    path "*.filtered.gtf", emit: gtf
+    path "versions.yml"  , emit: versions
+
+
```
Comment (drpatelh) [inline modules/cellranger/mkgtf/main.nf:26]:
```suggestion
```

File: modules/cellranger/count/main.nf [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    container "nfcore/cellranger:6.0.2"
+
+    input:
+    tuple val(meta), path(reads)
+    path reference
+
+    output:
+    path("sample-${meta.gem}/outs/*"), emit: outs
+    path "versions.yml"              , emit: versions
+
```
Comment (drpatelh) [inline modules/cellranger/count/main.nf:26]:
```suggestion
```

**Comment by drpatelh** [comment]
Need to get the tests running for these by adding to `pytest_modules.yml`?

**Comment by Emiller88** [comment]
> Need to get the tests running for these by adding to `pytest_modules.yml`?

I've played myself.

File: modules/cellranger/Dockerfile (hunk 1 above)
Comment (Emiller88) [reply modules/cellranger/Dockerfile:6]: On my TODO list to auto build and push any! Although this one is going to need to be ignored because of the 10x download :grimacing:

File: tests/config/pytest_modules.yml [hunk 14]
```diff
@@ -266,9 +266,19 @@ cat/fastq:
   - modules/cat/fastq/**
   - tests/modules/cat/fastq/**
 
-cellranger/mkref:
+cellranger/gtf: &cellranger/gtf
```
Comment (rpetit3) [inline tests/config/pytest_modules.yml:269]:
```suggestion
cellranger/gtf: &cellranger_gtf
```

Comment (rpetit3) [reply tests/config/pytest_modules.yml:269]: I think this should get the lint to pass

**Review by ggabernet — APPROVED** [review]
Looks good to me!
